NO-TICKET: Main Readme.md (#18)
iadvize/dataway: README.md

## Installation

```
npm install dataway
```

```
yarn add dataway
```

[API documentation](https://iadvize.github.io/dataway/docs)

## Introduction

`Dataway` is a datastructure representing the four possible states of a remote datasource fetching result.
* The remote datasource was NotAsked but eventually will be
* The remote datasource is Loading
* The remote datasource fetching has been a Success and a value was retrieved
* The remote datasource fetching ended up in a Failure and some error information was collected

> Using those 4 distinct states fix a classic issue where this is may be described in fewer step or more steps described with a bunch of boolean or worse detecting the presence of a value or catching an error, leading to unwanted states.

Reviewer: ```suggestion
This aims to solve a classic data management issue often handled either through booleans or complex, unwanted and polluted states. With one entry point and only 4 strongly-typed states, remote data handling becomes much cleaner.
```

Author: bien plus clair !

> Dataway also provide a great api, to manipulate, transform aggregate Dataway value in safe and optimistic way, actually reducing bug and crash occurence while making your code simpler to read

Reviewer: ```suggestion
Dataway also provides a great api to manipulate, transform and aggregate Dataway values in a safe and optimistic way. This reduces bug and crash occurence while making your code simpler to read
```

and more declarative.

## Example

> Imagine that our application rely on a webservice that provide us with a list of elements, and that our job is to both store the number of element in the application state for future usage, and display it.

Reviewer: ```suggestion
Imagine that our application relies on a webservice that provides us with a list of elements, and that our job is to both store the number of elements in the application state for future usage and to display it.
```

const view = state => {
  appElement.innerHTML = fold(
    () => "<p>Click on the load button</p>",
    () => "<p>Loading</p>",
    error => `<p>something wrong did happen : ${error}</p>`,
    success => `<ul>${success.map(post => `<li>${post.title}</li>`)}</ul>`,
    state
  );
};

const setState = stateManager(view, notAsked);

loadButton.onclick = event => {
  setState(loading);
  setTimeout(
    () =>
      fetch("https://jsonplaceholder.typicode.com/posts")
        .then(response => {
          if (response.ok) {
            return response.json();
          } else {
            return Promise.reject(
              `Request rejected with status ${response.status}`
            );
          }
        })
        .then(json => setState(success(json)))
        .catch(error => setState(failure(error))),
    2000
  );
};

```

## How to use

First you have to create some `Dataway` values using `notAsked`, `loading`, `failure(error)`, `success(value)`

```javascript
import { notAsked, failure, success } from 'dataway'

const foo = success('Mr Wilson');

const bar = failure('any suited error value');

const baz = notAsked;
```

> Then we can use the provided `map` api to apply a function on any `Success` variance of `Dataway`, wrapping automaticaly the result in a new `Success`.

Reviewer: ```suggestion
Then we can use the provided `map` api to apply a function on any `Success` variance of `Dataway`, wrapping automatically the result in a new `Success`.
```

> If provided variance of `Dataway` is not a `Success` it will be returned without change, and without executing the function.

Reviewer: ```suggestion
If the provided variance of `Dataway` is not a `Success`, it will be returned without change and without executing the function.
```

> This allow you as a developper to not have to check for `Dataway` variance before applying a function to its `Success` value.

Reviewer: ```suggestion
As a developper it means you do not have to check for `Dataway` variance before applying a function to its `Success` value.
```

```javascript
import { notAsked, failure, success, map } from 'dataway'

map(value => value.uppercase, success('Mr Wilson'));
// => Success "MR WILSON"

map(value => value.uppercase, failure('any suited error value'));
// => Failure "any suited error value"

map(value => value.uppercase, notAsked);
// => NotAsked
```

> Rewrapping the transformed value in a `Success` or returning the other variance untouched, allow 

Reviewer: ```suggestion
Rewrapping the transformed value in a `Success` or returning the other variance untouched, allows 
```

to transform a `Dataway` value in multiple distinct step wihout risking runtime error due to unexistant values (`null | undefined`) while keeping the variance of `Dataway` intact.

```javascript
import { notAsked, success, map } from 'dataway'

const upperCasedSuccess = map(value => value.uppercase, success('Mr Wilson'));
map(value => value.split(' '), upperCasedSuccess);
// => Success ['MR', 'WILSON']

const foo = map(value => value.uppercase, notAsked);
map(value => value.split(' '), foo);
// => NotAsked
```

To extract and use the `Success` value you must use the `fold` API.

> The following example illustrate how this force you to consider the four different UI to present to an user.

Reviewer: ```suggestion
The following example illustrates how this forces you to consider the four different UIs each state implies.
```

```javascript
import { success, map, fold } from 'dataway'

// => Success ['MR', 'WILSON']
const render = dataway => fold(
  () => "<p>Click on the load button</p>",
  () => "<p>Loading</p>",
  error => `<p>something wrong did happen : ${error}</p>`,
  success => `<p>${success}</p>`,
  dataway
);
render(success('Mr Wilson'));
// => <p>Mr Wilson</p>
render(failure('Ooops failed to fetch Mr Wilson data'));
// => <p>something wrong did happen : Ooops failed to fetch Mr Wilson data</p>
render(notAsked);
// => '<p>Click on the load button</p>'
render(loading);
// => '<p>Loading</p>'
```

> This is really great to simply create consistent UI for your users.

Reviewer: ```suggestion
This is really great to easily create consistent UIs.
```
UI -> forcément pour des users haha

> ## To continue

Reviewer: ```suggestion
## TLD;DR
```

> `Dataway` offer more api, to aggregate multiple dataway, to handle computation failure on your dataway value.

Reviewer: ```suggestion
`Dataway` offers a rich API api to aggregate multiple "dataways" or to handle computation failure on your dataway values.
```

> `Dataway` is written in typescript with toughtfull type description, allowing you to use it in typescript environnement without hassle while keeping great type safety from one end to another.

Reviewer: ```suggestion
`Dataway` is written in typescript with thoughtful type description, enabling you to use it in a typescript environnement without hassle while keeping great type safety.
```

> `Dataway` also offer compatibility with great libraries such as [Ramda](https://ramdajs.com), and [fp-ts](https://gcanti.github.io/fp-ts/)

Reviewer: ```suggestion
`Dataway` also offers compatibility with great libraries such as [Ramda](https://ramdajs.com), and [fp-ts](https://gcanti.github.io/fp-ts/)
```

> You can check and play with several example of usage 

Reviewer: ```suggestion
You can check and play with several examples
```
